### PR TITLE
Fix multiple memory bugs in toolbar customization

### PR DESCRIPTION
**Pull request by Technius**
This PR fixes multiple memory allocation bugs in the toolbar customizer.
* Fix double free when closing toolbar customization dialog.
* Fix separator icon not being rendered correctly when dragging to the toolbar.
* Fix errors involving the tool icons being left as floating references
* Refactor some of the toolbar customization code

There is a still a lingering issue where dragging & dropping a toolbar item may occasionally not trigger a redraw of the customization dialog.

### CODE REVIEW

File: src/gui/toolbarMenubar/model/ToolbarEntry.h
Comment (Febbe) [reply src/gui/toolbarMenubar/model/ToolbarEntry.h:22]: Forget that, it is not, the entries of the target list must be free'd, currently it is implemented with a leak (move). And behaves wrongly on self assign (copy).

**Review by Febbe — CHANGES_REQUESTED** [review]
**Future Todo:** We should probably implement a general GTK smart_pointer this will remove all problems with self assign, rule of 5, and self assignment bugs.
With that we could mark all constructors / destructors default.
Also, we could model the differences between floating and non-floating references.

File: src/gui/toolbarMenubar/model/ToolbarEntry.cpp
Comment (Febbe) [reply src/gui/toolbarMenubar/model/ToolbarEntry.cpp:55]: Maybe we get shorter lambdas in the future 😅.
Year,  c++ 20 adds something like `std::erase(container, item)` but we are still stuck on c++17 until apple decides to add enough of the stl to AppleClang.

I just hang there while reading your code and thought "wait, an iterator erase in a loop, ah puh, it's in an early exit block ok fine" 😬.

File: src/gui/toolbarMenubar/model/ToolbarEntry.cpp
Comment (Febbe) [reply src/gui/toolbarMenubar/model/ToolbarEntry.cpp]: Yes when you accidently move an reference of the object irrelevant from which place it came, on itself, you will clear the object and then you'll move an cleared object. This results in bugs which are extremely hard to find.

For reference google c++ self assignment.

Comment (Technius) [reply src/gui/toolbarMenubar/model/ToolbarEntry.cpp]: Well, that's not really a problem with the code but more like a problem with C++.

Comment (Febbe) [reply src/gui/toolbarMenubar/model/ToolbarEntry.cpp]: The thing is, that for both the copy and the move assignment operator override an existing Object, this means, that all special state must be cleared before override. In our case we must call `clearLists` before (move)assigning.
But this can result into bugs, when we use the same object as source and target:

```
Object k;
some_code_passing_k_als_reference(k);
some_other_code();


auto new_o = get_an_objectref_from_other_loc(); // This is surprisingly k
k = new_o;
```

After the assignment k will be an invalid wrong object.

Comment (Technius) [reply src/gui/toolbarMenubar/model/ToolbarEntry.cpp]: I see the problem now. I have fixed the error.

**Comment by Technius** [comment]
Thanks for the review, @Febbe. I think most of the PR is good to go. Not sure where the rendering issue is coming from though.

**Comment by idotobi** [comment]
@Technius : I've been working on the same files when implementig a configurable ColorPalette using .gpt files (https://github.com/xournalpp/xournalpp/pull/2379). I think your PR should be merged first as fixing bugs is more important than new features.
Please ping me once this one is merged so I can rebase and fix the merge-conflicts.

**Comment by Technius** [comment]
I'm going to merge this in 48 hours unless someone can provide a suggestion on how to fix the rendering issue.

**Comment by Febbe** [comment]
I could not reproduce the bug, does it only happen in your fix?

**Comment by idotobi** [comment]
> I'm going to merge this in 48 hours unless someone can provide a suggestion on how to fix the rendering issue.

:pray:  Thanks. But no hurry. I won't have time for the rebase before next weekend anyway :sweat_smile:

**Comment by Technius** [comment]
@Febbe Yes, this only occurs with the fix. Here's a demonstration:
xpp_toolbar_render

**Comment by idotobi** [comment]
@Technius : I tried to play around with the PR and for me the whole application crashes, but only for specific icons:

image

This is triggered by https://github.com/xournalpp/xournalpp/pull/2860/files#diff-580778d1227ae51eb0f9d5390970713285966d79d4ee11dd39dfe21a8ac1bd4bR144

For the ColorItems I fixed this in my PR by getting the PixBuf directly (https://github.com/xournalpp/xournalpp/pull/2379/files#diff-580778d1227ae51eb0f9d5390970713285966d79d4ee11dd39dfe21a8ac1bd4bR135) from `ColorSelectImage::newColorIconPixbuf` and not converting obtaining it from the `icon`. This seems to be broken also for 10.0.20 only there the assert is missing :sweat_smile:.

Maybe this would also help to resolve the other bug, which I wasn't able to reproduce :crossed_fingers:

**Comment by Technius** [comment]
Nice catch. The annoying part is that there is no uniform way to get image/icon data--a nice target for a refactor (out of scope of this PR).

**Comment by Febbe** [comment]
I'll do this, not that evil I think.

**Comment by Febbe** [comment]
It should work now, however some symbols are not visible in the toolbar-customize-dialog, and when dragging them from the dialog to the bar, they have a different symbol.

**Comment by idotobi** [comment]
I've been playing around a bit with this and there stills seems to be a few issues:

 - "Change Current Layer Name" has no icon even in toolbar

image

 - as @Febbe reported the icons change between toolbar and customization widget. The most notable is the printer Icon. Note that both during drag and on the toolbar the icon is the same. Only in the customization dialog it looks different. I'm not yet sure what to do with that insight though :thinking: 

Peek 2021-03-05 21-25

 - The ColorToolItems are double as big while dragging them from the Customization Dialog to the toolbar then vice versa. This one is easy to fix (I already did it in #2379). Once I'll figure out how to push to a PR directly :sweat_smile: 

Peek 2021-03-05 21-29

**Comment by Febbe** [comment]
> Once I'll figure out how to push to a PR directly

```
git remote add technius <his-git-url>

**Comment by idotobi** [comment]
Okay regarding my findings:

 - Change current Layer Name has no Icon

It seems like there was never an Icon for that. I also realized that the Draw/Select Combinations didn't have a good Icon representation so i created Icons for them. If you think they are better then missing/duplicate Icons I could include them.

image

 - Different Icons being used

This is simply due to different `GTK_ICON_SIZE` being used, which results in different Icons altogether for Stock Items. I would propose to get the size from the toolbar and then also use that size in the customize dialog. However, that will require some tinkering so I would not include this in this PR. Hardcoding the size could probably lead to a regression for some users.

 - Size of ColorToolItems during dragging has already been resolved

Annoyingly one cannot set the drag icon size for named icons. So for other icons, like the printer for example there would be no way to have a specific size.

**Comment by idotobi** [comment]
One more thing. The memory bug still occurs from time to time, but I cannot reproduce it systematically :see_no_evil:

**Comment by Febbe** [comment]
We will get new icons, so you should ask the designer if he likes them/if he can create some for those.

**Comment by idotobi** [comment]
@Febbe : Thanks for the pointer, that's definitely the better option.

**Comment by Technius** [comment]
Let's leave the "nice" icon problems to some other PR. This PR should focus only on the memory bugs.

The Mac build is failing due to a missing `_gtk_drag_set_icon_gicon` symbol. Any ideas?

Is the rendering issue still present?

**Comment by Febbe** [comment]
It's most likely, that the version of gtk we use from the blob, does not implement it.
Checkout: ToolbarDragDropHelper.cpp:25: there I set the Icon from a GIcon

**Comment by Technius** [comment]
The rendering issue is still present.

**Comment by idotobi** [comment]
@Technius : Did you find a way to systematically reproduce the bug? I still see the issue as well, but when I just tried to fix it I could not reproduce it neither for the changed code nor the original.

**Comment by Febbe** [comment]
gtk_drag_set_icon_gicon is not implemented for quartz: https://gitlab.gnome.org/GNOME/gtk/-/issues/3846
Therefore, we must remove the call, until it is implemented.

**Comment by idotobi** [comment]
While playing around with #2980  I realized that I don't see any change in behaviour during customization when I completely comment out the callback to `toolbarDragMotionCb`. 
Is this the same for you? Do you know what this function should actually do? 

If yes I would vote for either deleting it or at least deactivating it until it is fixed since it is called continuously during droping.

**Comment by Febbe** [comment]
@Technius I think you can rebase this PR

**Comment by Technius** [comment]
I rebased, squashing the fixup commits in the process. The Windows build failure seems to be a temporary configuration thing. If the PR looks fine, I'll merge in 24 hours.

**Comment by rolandlo** [comment]
https://github.com/xournalpp/xournalpp/blob/df20ccd1e66d11a796fd01661a2138d2bea353fb/src/gui/dialog/toolbarCustomize/ToolbarDragDropHelper.cpp#L22-L41

I'm confused. Isn't `gtk_drag_set_icon_gicon` still called here in the case `GTK_IMAGE_GICON` on MacOS and the precompiler statement only made in the (wrong) case `GTK_IMAGE_SURFACE`?

**Comment by Technius** [comment]
Oops, I made a mistake when rebasing. That ifndef guard block should go in the GICON part. Thanks for the catch.

**Pull request merged**
